### PR TITLE
file.py _get_flags needs to be reworked

### LINKED ISSUE #62676: [BUG] file.py _get_flags issue
**Description**
Running a simple salt file state results in failure. 
given /srv/salt/dnf-test.sls 

dnf_default_yes:
  file.append:
    - name: /etc/dnf/dnf.conf
    - text: 
      - defaultyes=True

**Setup**
Tested on a physical machine running Fedora 37, python 3.11, and salt v3005 (installed via the Fedora repo)

**Steps to Reproduce the behavior**
Create the following file:
```
/srv/salt/dnf-test.sls 

dnf_default_yes:
  file.append:
    - name: /etc/dnf/dnf.conf
    - text: 
      - defaultyes=True
```
Then run ```salt-call state.sls dnf-test```

This will result in the following error:
```
[ERROR   ] An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/salt/state.py", line 2276, in call
    ret = self.states[cdata["full"]](
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/salt/loader/lazy.py", line 149, in __call__
    return self.loader.run(run_func, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/salt/loader/lazy.py", line 1228, in run
    return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/salt/loader/lazy.py", line 1243, in _run_as
    return _func_or_method(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/salt/loader/lazy.py", line 1276, in wrapper
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/salt/states/file.py", line 6431, in append
    if __salt__["file.search"](
       ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/salt/loader/lazy.py", line 149, in __call__
    return self.loader.run(run_func, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/salt/loader/lazy.py", line 1228, in run
    return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/salt/loader/lazy.py", line 1243, in _run_as
    return _func_or_method(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/salt/modules/file.py", line 3141, in search
    return replace(
           ^^^^^^^^
  File "/usr/lib/python3.11/site-packages/salt/modules/file.py", line 2516, in replace
    flags_num = _get_flags(flags)
                ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/salt/modules/file.py", line 1643, in _get_flags
    _flag = getattr(re, str(flag).upper())
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 're' has no attribute 'RE.MULTILINE'
```

**Expected behavior**

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

```yaml
Salt Version:
          Salt: 3005
 
Dependency Versions:
          cffi: 1.15.1
      cherrypy: Not Installed
      dateutil: 2.8.2
     docker-py: Not Installed
         gitdb: 4.0.9
     gitpython: 3.1.27
        Jinja2: 3.0.3
       libgit2: 1.3.1
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 1.0.4
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.20
      pycrypto: Not Installed
  pycryptodome: 3.15.0
        pygit2: 1.7.1
        Python: 3.11.0rc1 (main, Aug  9 2022, 00:00:00) [GCC 12.1.1 20220628 (Red Hat 12.1.1-3)]
  python-gnupg: Not Installed
        PyYAML: 6.0
         PyZMQ: 23.2.0
         smmap: 5.0.0
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.3.4
 
System Versions:
          dist: fedora 37 Thirty Seven Prerelease
        locale: utf-8
       machine: x86_64
       release: 5.19.8-300.fc37.x86_64
        system: Linux
       version: Fedora Linux 37 Thirty Seven Prerelease

```
</details>

### PR DESCRIPTION
**Pull request by leifliddy**
### What does this PR do?
It causes the _get_flags function to return a bitwise integer as it was designed to do. 

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/62676

### Previous Behavior
The function ```_get_flags``` was designed to return an integer, but since python3.6 it returns the names of re flags instead
ie 
python 3.6 to 3.10
```
MULTILINE
# or
IGNORECASE|MULTILINE
```
and with python3.11, it returns this
```
re.MULTILINE
# or
re.IGNORECASE|re.MULTILINE
```
this proceeding ```re.``` causes an error with python3.11

### New Behavior
Rework this function so that in lieu of attempting to convert flag names to integers we instead convert (integers, strings, list of strings) to re flags. 
ie 
converting integer to re flags
```
elif isinstance(flags, int):
     return re.RegexFlag(flags)
```

get re flag name ie ```re.MULTILINE``` from flag name ie ```MULTILINE```
```
_flag = getattr(re.RegexFlag, str(flag).upper(), None)
```

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

### CODE REVIEW

**Comment by leifliddy** [comment]
Also, I didn't see a reason to use try/except: 
```            
        try:
            _flag = int(getattr(re, str(flag).upper()))
        except AttributeError as exc:
            raise CommandExecutionError("Exception: {}".format(exc))     
```
As the exception will be dumped to the screen anyway.

**Comment by OrangeDog** [comment]
> I didn't see any reason to use try/except:

To correctly wrap it so it can be handled as a Salt exception.

**Comment by OrangeDog** [comment]
This needs tests, and I'm pretty sure it doesn't actually fix the linked issue.

**Comment by leifliddy** [comment]
I know you're trying to help. But please let the devs handle this. Thanks.

**Comment by OrangeDog** [comment]
That's not a very friendly or collaborative attitude.

The linked issue appears to be because the passed flags are already flags, not strings.
The code is going to need to be something like this:

```python
_flags_acc = [0]  # reducing an empty list will raise TypeError
for flag in flags:
    if isinstance(flag, re.RegexFlag):
        _flags_acc.append(int(flag))
    else:
        _flag = getattr(re.RegexFlag, str(flag).upper(), None)
        if not isinstance(_flag, re.RegexFlag):
            raise SaltInvocationError(f"Invalid re flag given: {flag}")
        _flags_acc.append(int(_flag))
return reduce(operator.__or__, _flags_acc)
```

It can be simplified further by replacing the list with a single int, and `append` with `|=`.

**Comment by OrangeDog** [comment]
Converting the flags to `int` shouldn't actually be necessary either.

A better idea may be to always return a `RegexFlag` object. They're easily constructed:
```
>>> re.RegexFlag(24)
re.MULTILINE|re.DOTALL
```

**Comment by leifliddy** [comment]
That looks promising. I'm traveling ATM, but I'll play around with that logic late. Thanks.

**Comment by whytewolf** [comment]
> I know you're trying to help. But please let the devs handle this. Thanks.

@OrangeDog is right, this needs tests. And it also needs a changelog file. 
Also please be respectful. @OrangeDog helps out around here a lot. We cannot do everything. and cannot be in everywhere.

**Comment by leifliddy** [comment]
> ```python
> _flags_acc = [0]  # reducing an empty list will raise TypeError
> for flag in flags:
>     if isinstance(flag, re.RegexFlag):
>         _flags_acc.append(int(flag))
>     else:
>         _flag = getattr(re.RegexFlag, str(flag).upper(), None)
>         if not isinstance(_flag, re.RegexFlag):
>             raise SaltInvocationError(f"Invalid re flag given: {flag}")
>         _flags_acc.append(int(_flag))
> return reduce(operator.__or__, _flags_acc)
> ```

I really like this logic! I'm just trying to understand one piece of it. 
If this function is able to accept a ```string, a list of strings, or an integer``` then under what circumstances would this get hit on?

```python
if isinstance(flag, re.RegexFlag):
    _flags_acc.append(int(flag))
```

Oh wait I see...we 'd have to also allow the function to accept (and by extension return) a ```RegexFlag``` object as well. 
However, if this function accepted a ```RegexFlag``` why wouldn't we just return it (vs converting it to an integer and then converting back to a ```RegexFlag``` object? 

I mean that's what was done with an integer value before 
```python
elif isinstance(flags, int):
    return flags
```

Ok, I'll try and rework things to have the function accept and return a ```RegexFlag``` object. We'll see what that looks like...

**Comment by OrangeDog** [comment]
This was written back when regex flags were integers. In Python 3.6 they were changed into special enums, but this (and related) functions have not been updated before. Presumably there are some callers of this function that are passing flags, which now cause this problem when you try to look them up by their `__str__`.

As support for Python <3.6 is no longer needed, passing `int`s internally could be dropped, but the public interface still needs to handle them. That was the suggestion of my last comment:

```python
if instanceof(flags, re.RegexFlag):
  return flags
elif instanceof(flags, int):
  return re.RegexFlag(flags)
elif instanceof(flags, str):
  flags = [flags]

_flags = re.RegexFlag(0)
for flag in flags:
    if isinstance(flag, re.RegexFlag):
        _flags |= flag
    else:
        _flag = getattr(re.RegexFlag, str(flag).upper(), None)
        if not isinstance(_flag, re.RegexFlag):
            raise SaltInvocationError(f"Invalid re flag given: {flag}")
        _flags |= _flag
return _flags
```

But that probably also means changing (and thoroughly testing) the rest of the module that deals with regex flags.

**Comment by leifliddy** [comment]
I think this is the proper why to do it. I mean there's only a couple instances of where ```_add_flags``` and ```_get_flags``` are called. 

Just wondering whether we should edit this comment and allow ```replace``` to accept a ```re.RegexFlag``` data type.
```
flags (list or int)
    A list of flags defined in the ``re`` module documentation from the
    Python standard library. Each list item should be a string that will
    correlate to the human-friendly flag name. E.g., ``['IGNORECASE',
    'MULTILINE']``. Optionally, ``flags`` may be an int, with a value
    corresponding to the XOR (``|``) of all the desired flags. Defaults to
    8 (which supports 'MULTILINE').
```

Looking at your code, you're already checking if ```flags``` is an instance of ```re.RegexFlag```
```python
if instanceof(flags, re.RegexFlag):
  return flags
```
So why would you need that same check here? 
```python
for flag in flags:
    if isinstance(flag, re.RegexFlag):
        _flags |= flag
```

**Comment by OrangeDog** [comment]
There's no way for a user to give it a `re.RegexFlag` object. It has to be something deserializable from YAML.

**Comment by leifliddy** [comment]
There's something off with ```raise SaltInvocationError```
I mean you can literally do this 

```python
def _get_flags(flags):
    raise SaltInvocationError('xxxx')
```

And nothing will happen.   
Maybe there's a different/better salt exception we should be using in this function?

**Comment by leifliddy** [comment]
Uggg....trying to understand this function

```python
def search(path, pattern, flags=8, bufsize=1, ignore_if_missing=False, multiline=False):
 ...
    if multiline:
        print('is multi?')
        flags = _add_flags(flags, "MULTILINE")
        bufsize = "file"
```
I don't get why we're allowing ```flags``` and ```multiline``` to both be specified. 
One could easily cancel out the other. It's funny that the default value for flags is ```8``` (which is ```MULTILINE```), but the default value for ```multiline``` is ```false```
Maybe I'll just have it evaluate both variables for now....

Also I don't think ```bufsize``` is doing anything. I'm getting this error when enabling it.
````
/usr/lib/python3.11/site-packages/salt/utils/files.py:385: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used
  f_handle = open(*args, **kwargs)  # pylint: disable=resource-leakage
```

**Comment by OrangeDog** [comment]
`bufsize` is an even bigger job to sort out: #57584

**Comment by leifliddy** [comment]
@OrangeDog Ok I reworked ```file.py``` based on your suggestions. It could definitely be fined-tuned a bit especially with the leading comments in ```_get_flags``` and you've got a better python vocabulary then I do for sure. I like the idea of using the variable name ```re_flags``` (as opposed to just ```flags```) to capture the output of ```_get_flags``` and ```_add_flags``` and there's no mistaking what sort of data type that variable holds.

**Comment by leifliddy** [comment]
> 

Damn...

**Comment by leifliddy** [comment]
> `bufsize` is an even bigger job to sort out: #57584

Damn...

**Comment by leifliddy** [comment]
I'm not sure that this needs a test case for this as the module will continue to function as before. Only the data type that ```_get_flags``` returns has changed. I can't seem to find a single test case that verifies the data type returned by an internal function.

**Comment by OrangeDog** [comment]
The existing tests didn't catch the original issue, so there are more to be written there.

You can also write tests for the internal function. And/or ensure sufficient coverage of everything that calls it.

File: changelog/62676.fixed [hunk 1]
```diff
@@ -0,0 +1 @@
+Modified "_get_flags" function so that it returns regex flag names instead of integers
```
Comment (OrangeDog) [inline changelog/62676.fixed:1]: It doesn't return the names, it returns the flags.
The fix is to stop it failing when flags are provided as input.

Comment (leifliddy) [reply changelog/62676.fixed]: >The fix is to stop it failing when flags are provided as input.

Ok yes it was failing when ```re.MULTILINE``` was provided an input -- but it was getting that flag name from the ```_get_flags``` output (and was just feeding it back though via the ```_add_flags``` function). But sure we could verify all the input data types on that function. 
I've rarely written tests before, so this is going to be a bit of learning curve for me. I really just need to find a single test where someone's either verified the input and/or the output of an internal function...

**Comment by leifliddy** [comment]
This logic works 
```python
def test_get_flags_return():
    """
    Tests if flags are passed as the following data types
    integer, single flag string, list of strings, or as a regex flag
    """
    flags = 10
    ret = filemod._get_flags(flags)
    assert isinstance(ret, re.RegexFlag)
    
    flags = 'MULTILINE'
    ret = filemod._get_flags(flags)
    assert isinstance(ret, re.RegexFlag)
    
    flags = ['IGNORECASE', 'MULTILINE']
    ret = filemod._get_flags(flags)
    assert isinstance(ret, re.RegexFlag)
    
    flags = re.IGNORECASE|re.MULTILINE
    ret = filemod._get_flags(flags)
    assert isinstance(ret, re.RegexFlag)
```

Is everyone happy with that?

File: changelog/62676.fixed (hunk 1 above)
Comment (leifliddy) [reply changelog/62676.fixed]: I reworded this, thanks.

**Comment by leifliddy** [comment]
Screw it I'm going to change it to this so we can effectively verify the returned type and value
```python
def test__get_flags():
    """
    Test to ensure _get_flags returns a regex flag
    """
    flags = 10
    ret = filemod._get_flags(flags)
    assert ret == re.IGNORECASE | re.MULTILINE

    flags = "MULTILINE"
    ret = filemod._get_flags(flags)
    assert ret == re.MULTILINE

    flags = ["IGNORECASE", "MULTILINE"]
    ret = filemod._get_flags(flags)
    assert ret == re.IGNORECASE | re.MULTILINE

    flags = re.IGNORECASE | re.MULTILINE
    ret = filemod._get_flags(flags)
    assert ret == re.IGNORECASE | re.MULTILINE
```

Ok, I'm going to say this is done.

**Comment by leifliddy** [comment]
I've posted both a ```testcase``` and ```changelog``` file. How can get of those labels?

**Review by whytewolf — APPROVED** [review]

**Comment by leifliddy** [comment]
@whytewolf Shouldn't this PR have made it into ```v3005.1-2``` as it's a bugfix?

**Comment by whytewolf** [comment]
@leifliddy No, 3005.1-2 is package fixes only, things like requisites and python version for onedir and windows based installs. This is what that -2 means. it is still 3005.1 none of the code of salt changed.

**Comment by leifliddy** [comment]
Ahhhh... ok got it. Thanks for explaining that.

**Pull request merged**